python/src/iceberg/types.py

```diff
@@ -42,57 +42,63 @@ def __new__(cls, *args, **kwargs):
         return cls._instance
 
 
+@dataclass(frozen=True, eq=True, repr=True)
 class IcebergType:
-    """Base type for all Iceberg Types"""
-
-    _initialized = False
+    """Base type for all Iceberg Types
 
-    def __init__(self, type_string: str, repr_string: str):
-        self._type_string = type_string
-        self._repr_string = repr_string
-        self._initialized = True
+    Example:
+        >>> str(IcebergType())
+        'IcebergType()'
+        >>> repr(IcebergType())
+        'IcebergType()'
+    """
 
-    def __repr__(self):
-        return self._repr_string
+    type_string: str = field(init=False, repr=False)
 
     def __str__(self):
-        return self._type_string
+        if hasattr(self, "type_string"):
+            return self.type_string
+        return self.__repr__()
 
     @property
     def is_primitive(self) -> bool:
         return isinstance(self, PrimitiveType)
 
 
+@dataclass(frozen=True, eq=True)
 class PrimitiveType(IcebergType):
-    """Base class for all Iceberg Primitive Types"""
+    """Base class for all Iceberg Primitive Types
 
+    Example:
+        >>> str(PrimitiveType())
+        'PrimitiveType()'
+    """
 
+
+@dataclass(frozen=True, eq=True, repr=True)
```

Comment:
Aren't defaults for `eq` and `repr` `True`?  Is there a reason to be explicit about those two vs. relying on the defaults?

Reply:
The Python zen dictates: _Explicit is better than implicit_. But I agree that this might be too obvious. I'll remove eq and repr 👍🏻